pkg/kubelet/images/image_manager.go

```diff
@@ -224,3 +312,68 @@ func applyDefaultImageTag(image string) (string, error) {
 	}
 	return image, nil
 }
+
+// isEnsuredBySecret - returns two bools and error
+// 1. pulledBySecret: true if it is in ensured secret pulled image list,
+// the list will clean the image after PullImageSecretRecheckPeriod
+// 2. ensuredBySecret: true if the secret for an auth used to pull an
+// image has already been authenticated through a successful pull request
+// and the same auth exists for this podSandbox/image.
+func (m *imageManager) isEnsuredBySecret(imageRef string, image kubecontainer.ImageSpec, pullSecrets []v1.Secret) (bool, bool, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if imageRef == "" {
+		return false, false, errors.New("imageRef is empty")
+	}
+
+	// if the image is in the ensured secret pulled image list, it is pulled by secret
+	pulledBySecret := m.ensureSecretPulledImages[imageRef] != nil
+
+	img := image.Image
+	repoToPull, _, _, err := parsers.ParseImageName(img)
+	if err != nil {
+		return pulledBySecret, false, err
+	}
+
+	if m.keyring == nil {
+		return pulledBySecret, false, nil
+	}
+	keyring, err := credentialprovidersecrets.MakeDockerKeyring(pullSecrets, *m.keyring)
+	if err != nil {
+		return pulledBySecret, false, err
+	}
+
+	creds, withCredentials := keyring.Lookup(repoToPull)
+	if !withCredentials {
+		return pulledBySecret, false, nil
+	}
+	// if the pullImageSecretRecheckPeriod is 0, we don't need to check the secret again
+	if m.pullImageSecretRecheckPeriod.Duration == 0 {
+		return pulledBySecret, true, nil
+	}
+
+	for _, currentCreds := range creds {
+		auth := &runtimeapi.AuthConfig{
+			Username:      currentCreds.Username,
+			Password:      currentCreds.Password,
+			Auth:          currentCreds.Auth,
+			ServerAddress: currentCreds.ServerAddress,
+			IdentityToken: currentCreds.IdentityToken,
+			RegistryToken: currentCreds.RegistryToken,
+		}
+
+		hash, err := kubecontainer.HashAuth(auth)
+		if err != nil {
+			klog.ErrorS(err, "Failed to hash auth", "auth", auth)
+			continue
+		}
+		digest := m.ensureSecretPulledImages[imageRef]
+		if digest != nil {
+			ensuredInfo := digest.Auths[hash]
+			if ensuredInfo != nil && ensuredInfo.ensured && ensuredInfo.lastEnsuredDate.Add(m.pullImageSecretRecheckPeriod.Duration).After(time.Now()) {
```

Comment:
> If we have a boolean configuration for cross pod checking that can be set independently would we then default to some recheck period (cache invalidation) if not set or keep assuming image access for the entry in the cache till the kubelet is running (current in-memory impl)?

Not sure... I'd probably default both to current behavior and let admins opt into things that could break or disrupt currently working setups, but I hadn't thought about it for long.

Reply:
good point... let's add another config for the ensured by cross-pod checking option. I like the name.

What do you think about the default for the cross-pod checking option? Same as for the recheck period (defaults to off?)

Reply:
> What do you think about the default for the cross-pod checking option? Same as for the recheck period (defaults to off?)

Yeah, my starting position is *always* to default to settings that won't make an admin that failed to read 40 pages of release notes come hunt us down because we broke them on upgrade

Reply:
sounds good. We will have to document that setting the booelan and not the recheck period could have implications :)

Reply:
If recheck is disabled(with recheck duration is 0s), a new pod with same hash should be checked again.

It sounds a valid corner case.